File: transfer-lock-blocking-renewal/python/check_transfer_lock_risk.py
"""Detect a domain that is both transfer-locked and close to expiring.
Detection only. Removing a registrar transfer lock is an account or
registrar-portal action, not a Cloudflare DNS zone API call, so this
script never attempts a repair, it only reports the risky combination.
"""
import os
import logging
from datetime import datetime, timezone

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_transfer_lock_risk")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
WARNING_DAYS = int(os.environ.get("WARNING_DAYS", "30"))

LOCK_STATUSES = {"clienttransferprohibited", "servertransferprohibited"}


def assess_transfer_risk(status_codes: list, expiration_date: datetime, now: datetime, warning_days: int = 30) -> dict:
    """
    Pure decision logic, no I/O.
    Input: status_codes (raw RDAP/WHOIS status strings, any case/spacing),
           expiration_date (the domain's expiration as a datetime),
           now (the current time as a datetime, injected for testability),
           warning_days (how many days out counts as "near expiry").
    Output: {
        "locked": bool,
        "days_until_expiry": int,
        "at_risk": bool,
    }
    Logic: normalize status_codes to lowercase with spaces removed, and check
    membership against clienttransferprohibited/servertransferprohibited.
    Compute days_until_expiry as the whole number of days between now and
    expiration_date. at_risk is True only when locked is True AND
    days_until_expiry <= warning_days AND days_until_expiry >= 0.
    """
    normalized = {s.lower().replace(" ", "") for s in status_codes}
    locked = bool(normalized & LOCK_STATUSES)
    days_until_expiry = (expiration_date - now).days

    at_risk = locked and 0 <= days_until_expiry <= warning_days

    return {
        "locked": locked,
        "days_until_expiry": days_until_expiry,
        "at_risk": at_risk,
    }


def fetch_rdap(domain: str) -> dict:
    """Query RDAP for a domain and return the raw JSON response."""
    import requests

    r = requests.get(f"https://rdap.org/domain/{domain}", timeout=30)
    r.raise_for_status()
    return r.json()


def extract_expiration(rdap_data: dict) -> datetime:
    """Pull the expiration eventDate out of an RDAP response."""
    for event in rdap_data.get("events", []):
        if event.get("eventAction") == "expiration":
            iso = event["eventDate"].replace("Z", "+00:00")
            return datetime.fromisoformat(iso).astimezone(timezone.utc)
    raise ValueError("No expiration event found in RDAP response")


def report(domain: str, result: dict, status_codes: list) -> None:
    """Report the finding. Replace with email, Slack, or a webhook in production."""
    if not result["at_risk"]:
        log.info(
            "OK: %s locked=%s, %d day(s) until expiry, no action needed",
            domain, result["locked"], result["days_until_expiry"],
        )
        return
    prefix = "[dry run] would flag" if DRY_RUN else "FLAG"
    log.warning(
        "%s: %s is transfer-locked with only %d day(s) left before expiry. statuses=%s. "
        "Unlock at the registrar dashboard, or let auto-renew process first. "
        "This cannot be fixed through the Cloudflare DNS zone API.",
        prefix, domain, result["days_until_expiry"], status_codes,
    )


def run():
    log.info("Checking transfer lock risk for %s (DRY_RUN=%s)", DNS_DOMAIN, DRY_RUN)

    rdap_data = fetch_rdap(DNS_DOMAIN)
    status_codes = rdap_data.get("status", [])
    expiration_date = extract_expiration(rdap_data)
    now = datetime.now(timezone.utc)

    result = assess_transfer_risk(status_codes, expiration_date, now, WARNING_DAYS)
    report(DNS_DOMAIN, result, status_codes)


if __name__ == "__main__":
    run()

File: transfer-lock-blocking-renewal/node/check-transfer-lock-risk.js
/**
 * Detect a domain that is both transfer-locked and close to expiring.
 * Detection only. Removing a registrar transfer lock is an account or
 * registrar-portal action, not a Cloudflare DNS zone API call, so this
 * script never attempts a repair, it only reports the risky combination.
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const WARNING_DAYS = Number(process.env.WARNING_DAYS || 30);

const LOCK_STATUSES = new Set(["clienttransferprohibited", "servertransferprohibited"]);

/**
 * Pure decision logic, no I/O.
 * Input: statusCodes (raw RDAP/WHOIS status strings, any case/spacing),
 *        expirationDate (the domain's expiration as a Date),
 *        now (the current time as a Date, injected for testability),
 *        warningDays (how many days out counts as "near expiry").
 * Output: { locked: boolean, daysUntilExpiry: number, atRisk: boolean }
 * Logic: normalize statusCodes to lowercase with spaces removed, and check
 * membership against clienttransferprohibited/servertransferprohibited.
 * Compute daysUntilExpiry as the whole number of days between now and
 * expirationDate. atRisk is true only when locked is true AND
 * daysUntilExpiry <= warningDays AND daysUntilExpiry >= 0.
 */
export function assessTransferRisk(statusCodes, expirationDate, now, warningDays = 30) {
  const normalized = new Set(statusCodes.map((s) => s.toLowerCase().replace(/\s+/g, "")));
  const locked = [...LOCK_STATUSES].some((s) => normalized.has(s));
  const daysUntilExpiry = Math.floor((expirationDate.getTime() - now.getTime()) / 86400000);

  const atRisk = locked && daysUntilExpiry >= 0 && daysUntilExpiry <= warningDays;

  return { locked, daysUntilExpiry, atRisk };
}

/** Query RDAP for a domain and return the raw JSON response. */
async function fetchRdap(domain) {
  const res = await fetch(`https://rdap.org/domain/${domain}`);
  if (!res.ok) throw new Error(`RDAP returned ${res.status} for ${domain}`);
  return res.json();
}

/** Pull the expiration eventDate out of an RDAP response. */
function extractExpiration(rdapData) {
  const event = (rdapData.events || []).find((e) => e.eventAction === "expiration");
  if (!event) throw new Error("No expiration event found in RDAP response");
  return new Date(event.eventDate);
}

/** Report the finding. Replace with email, Slack, or a webhook in production. */
function report(domain, result, statusCodes) {
  if (!result.atRisk) {
    console.log(
      `OK: ${domain} locked=${result.locked}, ${result.daysUntilExpiry} day(s) until expiry, no action needed`
    );
    return;
  }
  const prefix = DRY_RUN ? "[dry run] would flag" : "FLAG";
  console.warn(
    `${prefix}: ${domain} is transfer-locked with only ${result.daysUntilExpiry} day(s) left before expiry. ` +
    `statuses=${JSON.stringify(statusCodes)}. Unlock at the registrar dashboard, or let auto-renew process first. ` +
    "This cannot be fixed through the Cloudflare DNS zone API."
  );
}

export async function run() {
  console.log(`Checking transfer lock risk for ${DNS_DOMAIN} (DRY_RUN=${DRY_RUN})`);

  const rdapData = await fetchRdap(DNS_DOMAIN);
  const statusCodes = rdapData.status || [];
  const expirationDate = extractExpiration(rdapData);
  const now = new Date();

  const result = assessTransferRisk(statusCodes, expirationDate, now, WARNING_DAYS);
  report(DNS_DOMAIN, result, statusCodes);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
